utils/zoneinfo64/src/lib.rs

```diff
@@ -117,54 +121,100 @@ impl<'a> ZoneInfo64<'a> {
 
-        #[expect(clippy::indexing_slicing)] // binary search
-        let name = self.names[idx];
-
-        #[expect(clippy::indexing_slicing)] // regions and names have the same length
-        let region = self.regions[idx];
-        #[expect(clippy::indexing_slicing)] // zones and names have the same length
-        let mut zone = &self.zones[idx];
-
-        #[expect(clippy::indexing_slicing)] // TzZone::Ints are validated as indices
-        if let &TzZone::Int(i) = zone {
-            zone = &self.zones[i as usize];
-        }
-        let TzZone::Table(ref zone) = zone else {
-            unreachable!() // data validate to have at most one alias jump
+        #[expect(clippy::indexing_slicing)] // just validated
+        let resolved_idx = if let TzZone::Int(i) = self.zones[idx] {
+            i as u16
+        } else {
+            idx as u16
         };
 
-        #[expect(clippy::indexing_slicing)] // rules indices are all valid
-        let final_rule = zone
-            .final_rule_offset_year
-            .map(|(idx, offset, year)| (&self.rules[idx as usize], offset, year))
-            .map(|(inner, offset_seconds, start_year)| Rule {
-                start_year,
-                standard_offset_seconds: offset_seconds,
-                inner,
-            });
-
+        // idx is a valid index into info.names by binary search
+        // zone_idx is a valid index into info.zones by the invariant that links don't point to links
         Some(Zone {
-            simple: zone.as_ref(),
-            final_rule,
-            region,
-            name,
+            idx: idx as u16,
+            resolved_idx,
+            info: self,
         })
     }
 }
 
-#[derive(Debug, Clone, Copy)]
+#[derive(Clone, Copy)]
 pub struct Zone<'a> {
-    simple: &'a TzZoneData<'a>,
-    final_rule: Option<Rule<'a>>,
-    pub name: &'a PotentialUtf16,
-    pub region: Region,
+    // a valid index into info.names
+    idx: u16,
+    // a resolved index into info.zones which points to a TzZone::Table
+    resolved_idx: u16,
+    info: &'a ZoneInfo64<'a>,
+}
+
+pub struct RawZone(u16);
+
+impl<'a> Zone<'a> {
+    /// Drops the internal reference to the [`ZoneInfo64`], eliminating the lifetime.
+    pub fn into_raw(self) -> RawZone {
+        RawZone(self.idx)
+    }
+
+    /// Reassociates the [`RawZone`] with a [`ZoneInfo64`].
+    ///
+    /// Returns garbage if the [`RawZone`] was obtained from a [`Zone`] that
+    /// was returned by a different [`ZoneInfo64`] than `info`.
+    pub fn from_raw_unchecked(RawZone(idx): RawZone, info: &'a ZoneInfo64<'a>) -> Self {
+        let idx = core::cmp::min(info.zones.len() - 1, idx as usize);
+        #[expect(clippy::indexing_slicing)]
+        // zones.len() is a valid index by the invariant that zones is non-empty, and everything
+        // below is trivially
+        let resolved_idx = if let TzZone::Int(i) = info.zones[idx] {
+            i as u16
+        } else {
+            idx as u16
+        };
```

Comment:
Issue: Document the function as panicky or make it return Option

Reply:
I'd prefer Option just for letting the caller decide

Reply:
there's even a comment justifying the allow

Reply:
Ah, should be fine then.